flink-connector-kudu/src/main/java/org/apache/flink/connector/kudu/source/enumerator/KuduSourceEnumerator.java

```diff
@@ -0,0 +1,264 @@
+        }
+    }
+
+    @Override
+    public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+        readersAwaitingSplit.add(subtaskId);
+        assignSplitsToReaders();
+    }
+
+    @Override
+    public void addSplitsBack(List<KuduSourceSplit> splits, int subtaskId) {
+        LOG.debug("Adding splits back: {}", splits);
+        pending.removeAll(splits);
+        unassigned.addAll(splits);
+        if (context.registeredReaders().containsKey(subtaskId)) {
+            readersAwaitingSplit.add(subtaskId);
+        }
+        assignSplitsToReaders();
+    }
+
+    @Override
+    public void addReader(int subtaskId) {
+        // The source is purely lazy-pull-based, nothing to do upon registration
+    }
+
+    @Override
+    public KuduSourceEnumeratorState snapshotState(long checkpointId) throws Exception {
+        return new KuduSourceEnumeratorState(lastEndTimestamp, unassigned, pending);
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            splitGenerator.close();
+        } catch (Exception e) {
+            throw new IOException("Error closing split generator", e);
+        }
+    }
+
+    @Override
+    public void handleSourceEvent(int subtaskId, SourceEvent sourceEvent) {
+        if (sourceEvent instanceof SplitFinishedEvent) {
+            SplitFinishedEvent splitFinishedEvent = (SplitFinishedEvent) sourceEvent;
+            LOG.debug(
+                    "Received SplitFinishedEvent from subtask {} for splits: {}",
+                    subtaskId,
+                    splitFinishedEvent.getFinishedSplits());
+            pending.removeAll(splitFinishedEvent.getFinishedSplits());
+            readersAwaitingSplit.add(subtaskId);
+            assignSplitsToReaders();
+        }
+    }
+
+    // This function is invoked repeatedly according to this.period if there are no outstanding
+    // splits.
+    // Outstanding meaning that there are no pending splits, and no enumerated but not assigned
+    // splits for the
+    // current period.
+    private List<KuduSourceSplit> enumerateNewSplits(Supplier<Boolean> shouldGenerate) {
+        if (!shouldGenerate.get()) {
+            return null;
+        }
+        List<KuduSourceSplit> newSplits;
+
+        if (isFirstSplitGeneration()) {
+            lastEndTimestamp = KuduSourceUtils.getCurrentHybridTime();
+            newSplits = splitGenerator.generateFullScanSplits(lastEndTimestamp);
+        } else {
+            long startHT = lastEndTimestamp;
+            long endHT = KuduSourceUtils.getCurrentHybridTime();
+            newSplits = splitGenerator.generateIncrementalSplits(startHT, endHT);
+            lastEndTimestamp = endHT;
+        }
+
+        return newSplits;
+    }
+
+    private void assignSplits(List<KuduSourceSplit> splits, Throwable error) {
+        if (error != null) {
+            LOG.error("Failed to enumerate Kudu splits.", error);
+            return;
```

Comment:
Instead of logging, we should throw a `FlinkRuntimeException` here, cause otherwise for example if the Kudu instance becomes unreachable the Flink job will keep running logging the same exception again and again in every discovery period. We should also shut down the Flink job if the source is unreachable. The recovery configuration will try to redeploy the Flink job if it matters anyways and then it will be obvious there is a problem.